Error when attempting to retrieve Cloudflare Load Balancer Pool configuration in Terraform using cf-terraforming

Cannot get an export of the load balancer pools in Terraform format. Error as per below,

panic: interface conversion: interface {} is []interface {}, not map[string]interface {}
goroutine 1 [running]:
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.nestBlocks(0xc000756680, 0xc0004c7ad0, 0x2, 0x0, 0x0)
/cf-terraforming/internal/app/cf-terraforming/cmd/util.go:187 +0x135f
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.generateResources.func1(0x18f8f20, 0xc0001e1d40, 0x0, 0x4)
/cf-terraforming/internal/app/cf-terraforming/cmd/generate.go:645 +0x1059
github.com/spf13/cobra.(*Command).execute(0x18f8f20, 0xc0001e1d00, 0x4, 0x4, 0x18f8f20, 0xc0001e1d00)
/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x18f9420, 0xc00027ff78, 0x4080c5, 0xc00008a058)
/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.Execute()
/cf-terraforming/internal/app/cf-terraforming/cmd/root.go:30 +0x31
main.main()
/cf-terraforming/cmd/cf-terraforming/main.go:8 +0x25

I am using Docker Desktop for Windows. This is the Docker command I am running,

docker run -it --env CLOUDFLARE_EMAIL=<email_address_here>--env CLOUDFLARE_API_KEY=<global_api_key_here> --env CLOUDFLARE_ACCOUNT_ID=<CF_accountid_here> -v c:/docker:/cloudflare --workdir=/cloudflare cfterraforming generate --resource-type cloudflare_load_balancer_pool --zone <CF_zoneid_here> >> cloudflare.tf

An example, I am able to run the same command above and replace cloudflare_load_balancer_pool with cloudflare_load_balancer_monitor which works fine and I can retrieve the load balancer monitors.

The Docker file I used to create the image is as per below,

FROM golang:1.15-alpine AS build
ENV GO111MODULE=on
RUN apk add git \
&& git clone https://github.com/cloudflare/cf-terraforming.git /cf-terraforming \
&& cd /cf-terraforming \
&& go build -o cf-terraforming cmd/cf-terraforming/main.go
FROM alpine:3
RUN apk add ca-certificates && update-ca-certificates
COPY --from=build /cf-terraforming/cf-terraforming /usr/local/bin/cf-terraforming
ENTRYPOINT ["/usr/local/bin/cf-terraforming"]

Using the Docker image created from the file I am able to retrieve different resource configurations in Terraform format usinf cf-terraforming. It seems, for now, it is just the cloudflare_load_balancer_pool which is giving an error.

Hi,

I tried golang 1.16.5-alpine and alpine 3.13. I am assuming the latest version of cf-terraforming will be utilized, are you able to confirm? Got a similar error, maybe the same error, when trying to retrieve the load balancer pool configuration but for other resources e.g. dns records & load balancer monitors it works ok.

Thanks.

Docker file as per below,

FROM golang:1.16.5-alpine AS build
ENV GO111MODULE=on

RUN apk add git \
    && git clone https://github.com/cloudflare/cf-terraforming.git /cf-terraforming \
    && cd /cf-terraforming \
    && go build -o cf-terraforming cmd/cf-terraforming/main.go
	
FROM alpine:3.13
RUN apk add ca-certificates && update-ca-certificates
COPY --from=build /cf-terraforming/cf-terraforming /usr/local/bin/cf-terraforming
ENTRYPOINT ["/usr/local/bin/cf-terraforming"]

Error as per below,

panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 1 [running]:
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.nestBlocks(0xc0005cd400, 0xc00065bbc0, 0x2, 0xc00003e5d0, 0x30)
	/cf-terraforming/internal/app/cf-terraforming/cmd/util.go:187 +0x131f
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.generateResources.func1(0x14c0de0, 0xc000185c40, 0x0, 0x4)
	/cf-terraforming/internal/app/cf-terraforming/cmd/generate.go:645 +0x1059
github.com/spf13/cobra.(*Command).execute(0x14c0de0, 0xc000185c00, 0x4, 0x4, 0x14c0de0, 0xc000185c00)
	/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x14c12e0, 0xc0002dff78, 0x4080e5, 0xc00008a058)
	/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.Execute()
	/cf-terraforming/internal/app/cf-terraforming/cmd/root.go:30 +0x31
main.main()
	/cf-terraforming/cmd/cf-terraforming/main.go:8 +0x25

Hi,

I've got cf-terraforming working in a cmd prompt on Windows. The documentation available online to get it working is not very clear. Anyways, I'm using the latest version now. cf-terraforming version details as per below,

C:\cf-terraforming> .\cf-terraforming.exe version
-
cf-terraforming v0.3.0

I still get an error when trying to get an export of the load balancer pools. I have tried other resources e.g. dns records and load balancer monitors and that works ok.

Thanks.

Error as per below,

time="2021-06-10T20:21:54+10:00" level=debug msg="initializing cloudflare-go" account_id=<account_id_displayed_here> email=<email_displayed_here>  zone_id=
time="2021-06-10T20:21:54+10:00" level=debug msg="configuring Cloudflare API with account" account_id=<account_id_displayed_here>
time="2021-06-10T20:23:01+10:00" level=debug msg="initialising Terraform in ."
time="2021-06-10T20:23:20+10:00" level=debug msg="reading Terraform schema for Cloudflare provider"
time="2021-06-10T20:23:21+10:00" level=debug msg="beginning to read and build cloudflare_load_balancer_pool resources"
panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 1 [running]:
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.nestBlocks(0xc000234e00, 0xc000665830, 0x2, 0xc000577890, 0x30)
        /home/runner/work/cf-terraforming/cf-terraforming/internal/app/cf-terraforming/cmd/util.go:187 +0x131f
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.generateResources.func1(0x14ff940, 0xc00044a2d0, 0x0, 0x9)
        /home/runner/work/cf-terraforming/cf-terraforming/internal/app/cf-terraforming/cmd/generate.go:644 +0xfe5
github.com/spf13/cobra.(*Command).execute(0x14ff940, 0xc00044a240, 0x9, 0x9, 0x14ff940, 0xc00044a240)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x14ffe40, 0xc0002dff78, 0x41670b, 0xc00003a0b8)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.Execute()
        /home/runner/work/cf-terraforming/cf-terraforming/internal/app/cf-terraforming/cmd/root.go:30 +0x38
main.main()
        /home/runner/work/cf-terraforming/cf-terraforming/cmd/cf-terraforming/main.go:8 +0x27

Still not working correctly :-(

```
$ cf-terraforming generate --resource-type "cloudflare_load_balancer_pool" --account $CLOUDFLARE_ACCOUNT_ID -z $CLOUDFLARE_ZONE_ID > cloudflare_load_balancer_pool.tf
panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 1 [running]:
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.nestBlocks(0xc000549b80, 0xc001994a20, 0x2, 0xc0012cf7a0, 0x30)
	/home/runner/work/cf-terraforming/cf-terraforming/internal/app/cf-terraforming/cmd/util.go:187 +0x131f
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.generateResources.func1(0x14c4e20, 0xc00050c180, 0x0, 0x6)
	/home/runner/work/cf-terraforming/cf-terraforming/internal/app/cf-terraforming/cmd/generate.go:644 +0xfc5
github.com/spf13/cobra.(*Command).execute(0x14c4e20, 0xc00050c120, 0x6, 0x6, 0x14c4e20, 0xc00050c120)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x14c5320, 0xc0002dff78, 0x407fe5, 0xc000038178)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/cloudflare/cf-terraforming/internal/app/cf-terraforming/cmd.Execute()
	/home/runner/work/cf-terraforming/cf-terraforming/internal/app/cf-terraforming/cmd/root.go:30 +0x31
main.main()
	/home/runner/work/cf-terraforming/cf-terraforming/cmd/cf-terraforming/main.go:8 +0x25
```

--- a/internal/app/cf-terraforming/cmd/generate.go
+++ b/internal/app/cf-terraforming/cmd/generate.go
@@ -417,6 +417,15 @@ func generateResources() func(cmd *cobra.Command, args []string) {
 			resourceCount = len(jsonPayload)
 			m, _ := json.Marshal(jsonPayload)
 			json.Unmarshal(m, &jsonStructData)
+
+			for i := 0; i < resourceCount; i++ {
+				for originCounter, _ := range jsonStructData[i].(map[string]interface{})["origins"].([]interface{}) {
+					if jsonStructData[i].(map[string]interface{})["origins"].([]interface{})[originCounter].(map[string]interface{})["header"] != nil {
+						jsonStructData[i].(map[string]interface{})["origins"].([]interface{})[originCounter].(map[string]interface{})["header"].(map[string]interface{})["header"] = "Host"
+						jsonStructData[i].(map[string]interface{})["origins"].([]interface{})[originCounter].(map[string]interface{})["header"].(map[string]interface{})["values"] = jsonStructData[i].(map[string]interface{})["origins"].([]interface{})[originCounter].(map[string]interface{})["header"].(map[string]interface{})["Host"]
+					}
+				}
+			}
 		case "cloudflare_load_balancer_monitor":
 			jsonPayload, err := api.ListLoadBalancerMonitors(context.Background())
 			if err != nil {

--- a/internal/app/cf-terraforming/cmd/generate_test.go
+++ b/internal/app/cf-terraforming/cmd/generate_test.go
@@ -100,6 +100,7 @@ func TestResourceGeneration(t *testing.T) {
 		"cloudflare health check":                           {identiferType: "zone", resourceType: "cloudflare_healthcheck", testdataFilename: "cloudflare_healthcheck"},
 		"cloudflare load balancer monitor":                  {identiferType: "account", resourceType: "cloudflare_load_balancer_monitor", testdataFilename: "cloudflare_load_balancer_monitor"},
 		"cloudflare load balancer":                          {identiferType: "zone", resourceType: "cloudflare_load_balancer", testdataFilename: "cloudflare_load_balancer"},
+		"cloudflare load balancer pool":                     {identiferType: "account", resourceType: "cloudflare_load_balancer_pool", testdataFilename: "cloudflare_load_balancer_pool"},
 		"cloudflare logpush jobs":                           {identiferType: "zone", resourceType: "cloudflare_logpush_job", testdataFilename: "cloudflare_logpush_job"},
 		"cloudflare origin CA certificate":                  {identiferType: "zone", resourceType: "cloudflare_origin_ca_certificate", testdataFilename: "cloudflare_origin_ca_certificate"},
 		"cloudflare page rule":                              {identiferType: "zone", resourceType: "cloudflare_page_rule", testdataFilename: "cloudflare_page_rule"},

--- a/testdata/terraform/cloudflare_load_balancer_pool.tf
+++ b/testdata/terraform/cloudflare_load_balancer_pool.tf
@@ -6,4 +6,14 @@ resource "cloudflare_load_balancer_pool" "terraform_managed_resource" {
   monitor            = "f1aba936b94213e5b8dca0c0dbf1f9cc"
   name               = "primary-dc-1"
   notification_email = "someone@example.com,sometwo@example.com"
+  origins {
+    address = "0.0.0.0"
+    enabled = true
+    name    = "app-server-1"
+    weight  = 1
+    header {
+      header = "Host"
+      values = ["example.com"]
+    }
+  }
 }

